### PR TITLE
FIX: Repeated iteration should return all records

### PR DESCRIPTION
**Pull request by sminnee**
See https://github.com/silverstripe/silverstripe-framework/issues/9097

This is a bugfix whose root-cause is a long-standing misunderstanding of the Iterator::rewind() API. Therefore this change is correcting that – rewind() no longer returns a value; nor does its supporting method, seek().

Is this an API change or a bugfix? Given that this is generally speaking an internal API accessed via Iterator, and our implementation is wrong, I recommend that we include it on the 4.4 branch. To put it another way, the stated API _was_ the bug.

### CODE REVIEW

**Comment by sminnee** [comment]
To put it another way - what's more likely: that people are going to be iterating on queries more than once, or calling and that they are calling seek() and rewind() directly? I think the former.

Arguably, I could keep PDOQuery::rewind() and seek() returning a value because of the memory-inefficient way it has been built. Happy to get feedback on that.

**Comment by sminnee** [comment]
The other thing I could do is preserve the behaviour of seek() by

 * Defining `seekWithoutReturn()` on MySQLQuery and PostgreSQLQuery, marked experimental.
 * Calling that if it exists in favour of `seek()` (but not including it as abstract in Query)

That, and excluding our mis-implementation of Iterator::rewind() from the public API, would mean that we're not breaking a public API with this change. Bit messier code, but that seems okay. I'd put a simpler fix on `master`  too.

**Comment by ScopeyNZ** [comment]
Can't we just fix this by:

```
    public function seek($row)
    {
        if (is_object($this->handle)) {
            $this->handle->data_seek($row);
            $record = $this->nextRecord();
            $this->handle->data_seek($row);
            return $record;
        }
        return null;
    }
```

Yes it's weird but it's non-breaking. And hopefully the cost of seeking is negligible in this case?

**Comment by sminnee** [comment]
> And hopefully the cost of seeking is negligible in this case?

Doubtful.

**Comment by sminnee** [comment]
But we should probably do some profiling.

**Comment by ScopeyNZ** [comment]
I don't really think this API is well used. I'd much rather accept the patch as-is rather than introducing a `seekWithoutReturn` API. If the seek cost is measurable then I think it's also worth accepting this patch as is - assuming that rewind is sometimes called without it actually being necessarily required (more than it's called and then used).

**Comment by ScopeyNZ** [comment]
Also, I'm okay with dropping the return on `rewind` (that's Iterator API definition that we're breaking). `seek` is our own API definition though.

**Comment by sminnee** [comment]
OK well your change doubles the execution time of calling map() on a 4-records resultset 100 times.

However, for a 100 executions it goes from 0.55ms to 1.1ms. So it's probably negligible, especially given that this is a somewhat rare API. The proportionate difference was also less with 10,000 executions (74ms vs 99ms).

So I'll put Guy's suggestion in the PR.

**Comment by sminnee** [comment]
This PR will be needed to get non-PDO PGSQL passing

**Comment by sminnee** [comment]
It looks like filterByCallback at least is returning the first record twice, so we're erroring the other way! :-| 

TLDR: this still needs a bit of work

**Comment by sminnee** [comment]
OK looks like the fix to MySQLStatement::rewind() has sorted this out; we'll just need a fix for PGSQL (and a linting fix)

**Comment by sminnee** [comment]
I'm also gonna rebase this against 4 so that the php 7.4 test suite is reinstated.

File: src/ORM/Connect/MySQLStatement.php [hunk 1]
```diff
@@ -115,6 +115,7 @@ public function seek($row)
 
     public function numRecords()
     {
+        echo " - numRows()\n";
```
Comment (ScopeyNZ) [inline src/ORM/Connect/MySQLStatement.php]: Whoops

**Comment by sminnee** [comment]
OK relevant postgresql fix has been merged now and I've pushed a linting fix. I think this is ready to merge if the tests pass, but there shouldn't be any "benign" test failure now.

**Comment by sminnee** [comment]
💚
